### ISSUE TITLE
Tinkering with another way of computing metrics

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/ocf-ml-metrics#11]
Hiya,

Thanks so much for creating this `ocf-ml-metrics` code base! It's super-useful!

Over in the `gb_national_pv_forecast` repo, I'm tinkering with a way to make the metrics composable (inspired by your work on data pipes :slightly_smiling_face:). I'm starting with just a single location at a time (because I'm focused on _national_ PV forecasting!)

One aim is to completely separate the different steps of the "metrics pipeline": To separate the denormalization, ignoring night, aggregating by different time frequencies, etc.

Below is the initial draft design... my guess is that I should flesh this out a bit in the `gb_national_pv_forecast` repo, to see if this design actually works well in practice. (My ultimate aim is to have a set of metrics that the code can compute on every epoch, both for the training and testing batches. And send the results to Weights and Biases, perhaps as a `wandb.Table`). And, if it does, then we can consider merging this "composable" approach into `ocf-ml-metrics`. But absolutely no pressure! I'm just tinkering! And this design might turn out to be more pain than it's worth...

## Initial design

The metrics functions expect `predictions` to be a `pandas.DataFrame`:

* Index: `target_datetime_utc`
* Columns:
  - `t0_datetime_utc`
  - `forecast_pv_yield`  # Values must be in the range [0, 1]

The `actual` values are a `pd.Series`, where the index is the `target_datetime_utc`,
and the values are the `actual_pv_yield` (in the range [0, 1]).

Metrics are defined by a `metrics_pipeline` which is a tuple of one or more `callable` objects.
For example:

```python
# Define re-usable metrics pipelines: The dict key is a human-readable metric name,
# and the dict value is the pipeline itself.
metrics_pipelines: dict[str, tuple[Callable, ...]] = {
    "MAE in MW, ignoring night": (
        Denormalize(national_pv_capacity_mwp_per_target_datetime_utc),
        IgnoreNight(latitude, longitude, threshold_for_sun_angle_in_degrees=-5),
        absolute_error,
        mean,
    ),
    "NMAE per month, including night": (
        absolute_error,
        GroupBy(pd.Grouper(freq="M")),
        mean,
    ),
    "NMAE per hour, ignoring night": (
        IgnoreNight(latitude, longitude, threshold_for_sun_angle_in_degrees=-5),
        absolute_error,
        GroupBy(pd.Grouper(freq="H")),
        mean,
    )
}

# Run the pipelines:
metrics_results: dict[str, Union[Number, pd.Series, pd.DataFrame]] = run_all_pipelines(
    metrics_pipelines, predictions, actual)
```

**Comment by jacobbieker** [comment]
Yeah, I think making it more like the datapipes makes sense to me! Would be interesting to see how easy it is to work with, but I do like the datapipes design, so I think it probably would be helpful here too.
